### PR TITLE
silence "no symbols" warnings on apple

### PR DESCRIPTION
**Pull request by andresag01**
## Description
This change fixes the "no symbols" warning when building the .ar files with both CMake and Makefile under Mac OS X. This PR superseeds https://github.com/ARMmbed/mbedtls/pull/1600 and adds a ChangeLog entry and ports the change to Makefile.

## Status
**IN DEVELOPMENT**

## Requires Backporting
Yes
Mbedtls 2.1:
Mbedtls 2.7:

## Additional comments
The change adds an APPLE_BUILD (and APPLE) option for Makefiles. Use the following command to build Mbed TLS on OS X:
```
make APPLE=1
```
or
```
make APPLE_BUILD=1
```
Note that having both variables `APPLE` and `APPLE_BUILD` is analogous to the `WINDOWS` and `WINDOWS_BUILD` system, so I have added both to be consistent (though one is redundant).

## Todos
- [ ] Tests: I tested the change on Linux and Mac, but it would be good to check that the makefiles still work under Windows. Also, I think the CMake change does not take into account setting the `-` before the options to the archiver as it is the case in make. I am not sure this is a problem though...

### CODE REVIEW

**Comment by andresag01** [comment]
@Patater: What command did you run to build the library? It builds without comments in my system. Please note that I have updated the PR description to explain that you need to set the `APPLE` make variable for it to work (as in `make APPLE=1`).

>And I don't get warnings from the RL step anymore, but it looks like the Makefile changes cause the RL step not to get run anymore. Is APPLE_BUILD correct?

I modified the Makefile so that on Linux the `ar` and `ranlib` (RL) steps are executed in a single invocation or `ar`, thats why if APPLE=1 is not set, then RL is not printed. However, please note that the output archive should be the same in Linux as before the patch because `ar` is executed with the `-s` flag.

When APPLE build is set in make, then the `ar` and `ranlib` steps are split because some additional flags need to be passed to `ranlib` to not emit warnings when some object files do not have symbols. Note that this warning occurs when for example a module is disable, so it has no functions but it is still "compiled". Then `ranlib` finds that there are no symbols for some object files and issues a warnings in OS X.

**Comment by andresag01** [comment]
@Patater: Given that there is a new build option (APPLE) I will wait for the changes to be approved before backporting.

**Comment by Patater** [comment]
I expected `APPLE_BUILD` or `APPLE` to be automatically set, so I just ran `make`. It's a bad experience to have to manually specify the OS. It should be possible to do something like `ifeq ($(shell uname -s),Darwin)` from the Makefile and set APPLE_BUILD automatically (with `?=` I suppose).

**Comment by andresag01** [comment]
@Patater: I like the idea that the build system will autodetect the underlying OS is apple, but I did not add this because it was not done for Windows. Should I then add the autodetection?

**Comment by andresag01** [comment]
@Patater: I have now added autodetection of OS X in the library/Makefile and removed the APPLE option.

File: library/Makefile [hunk 1]
```diff
@@ -21,6 +21,8 @@ endif
 # if were running on Windows build for Windows
 ifdef WINDOWS
 WINDOWS_BUILD=1
+else ifeq ($(shell uname -s),Darwin)
+APPLE_BUILD=1
```
Comment (Patater) [inline library/Makefile]: Should we use `?=`? Our Makefiles may not support cross-compiling, but if they do, then we may want to allow the user to force APPLE_BUILD to 0 even when their host OS is macOS. There may also be other reasons to use `?=` besides the cross compiling case.

**Comment by andresag01** [comment]
@Patater: Reworked PR according to your comments.

File: library/Makefile [hunk 2]
```diff
@@ -35,16 +37,26 @@ SOEXT_TLS=so.10
...
-# Set DLEXT=dylib to compile as a shared library for Mac OS X
 DLEXT ?= so
 
 # Set AR_DASH= (empty string) to use an ar implentation that does not accept
 # the - prefix for command line options (e.g. llvm-ar)
 AR_DASH ?= -
 
-# Windows shared library extension:
+ifdef APPLE_BUILD
+ARFLAGS = $(AR_DASH)Src
```
Comment (mazimkhan) [inline library/Makefile]: Why are these flags dependent on the platform and platform alone. These flags can change with compiler. If they change for a platform  and compiler selection then both should be checked. Have tested that this change will suit all compilers?

Comment (Patater) [reply library/Makefile]: Due to peculiarities of macOS, it's really rare to use the non-system compiler. It's possible, but requires going out to Mac Homebrew or other 3rd party packages, or building the compiler yourself from source. Due to how rare using other compilers is, we should deal with any issues in using those compilers with our Makefile as separate bugs and PRs to fix those bugs.

File: library/Makefile [hunk 3]
```diff
@@ -21,6 +21,8 @@ endif
 # if were running on Windows build for Windows
 ifdef WINDOWS
 WINDOWS_BUILD=1
+else ifeq ($(shell uname -s),Darwin)
```
Comment (sbutcher-arm) [inline library/Makefile]: Is this going to work for cross-compilation? Maybe we should qualify the test with a check that the compiler is cc, and not `cc` or `clang`?

**Comment by andresag01** [comment]
@sbutcher-arm: Building with Makefiles for OS X should work without any issues now. Cross compilation on OS X also works, but because of the autodetection of the underlying system that we have added, we need to tell the makefile to not alter the flags to ar and ranlib. So, make must be executed with APPLE_BUILD=0 in that case, for example:
```
make APPLE_BUILD=0 CC=arm-none-eabi-gcc AR=arm-none-eabi-ar LD=arm-none-eabi-ld lib
```
Is that acceptable? Alternatively, a fix for this cross compilation issue would be your suggestion to check the contents of AR, LD and CC, but I that seems fragile to me, so I would prefer to avoid it.

Please note that as far as I can tell, the cross compilation issue is not a problem for our tests. I think all.sh is the only place where such tests are run, but that works only on Linux. The changes have no usage impact on any system except cross compiling on Mac OS X

**Comment by andresag01** [comment]
@Patater @sbutcher-arm: To avoid the problem of having to define APPLE_BUILD=0 manually in the commandline when cross compiling, I added a check to ensure that the apple-specific arguments to the archiver are only set when we are using the system archiver. To do that, the makefile first checks whether it is running under Mac OS X, if that succeeds then it checks if `$(AR) == "ar"` and if both checks succeed, then APPLE_BUILD is set to 1. In the other cases, (e.g. running under windows, running under linux, cross compiling) the makefiles are expected to behave as it was before without the need of setting any flags.

File: library/CMakeLists.txt [hunk 4]
```diff
@@ -96,6 +96,13 @@ if(WIN32)
     set(libs ${libs} ws2_32)
 endif(WIN32)
 
+if(APPLE)
```
Comment (sbutcher-arm) [inline library/CMakeLists.txt]: At the moment I have to specify to CMake that this is an Apple build with `cmake -DAPPLE=On .`. That shouldn't be necessary, and I think it should autodetect OSX. Googling the problem, I think you should have platform detection, something like:
```
string(REGEX MATCH "Darwin" CMAKE_PLATFORM_IS_OSX "${CMAKE_SYSTEM_NAME}")
```
And then use `CMAKE_PLATFORM_IS_OSX` in place of `APPLE`.

Our CMake files can't cross compile unless you specify IAR as the compiler, so there should be no similar issue as with the GNU Makefiles.

**Comment by andresag01** [comment]
@sbutcher-arm: Just to clarify:

* No further changes/adjustments are necessary for the Makefile builds.
* The CMake builds do not correctly support cross-compilation, but this is a preexisting problem. This problem will be addressed in a separate ticket.
* CMake builds do not automatically detect they are running on OS X.
    * Last time I tried this on OS X it was working without having to specify `-DAPPLE=On`. Also, my understanding from the [CMake docs](https://cmake.org/cmake/help/v3.6/variable/APPLE.html) is that the `APPLE` variable is set by CMake, so there should not be a need to set that on the command line argument to CMake. I will take a look again.
    * If CMake does automatically set `APPLE` then there might be issues with the archiver once again under OS X when cross compiling. Should we leave that to be fix when the cross compilation actually works on that build system?

**Comment by sbutcher-arm** [comment]
@andresag01 - When I looked into if it was possible to detect OS X, inspecting `CMAKE_SYSTEM_NAME ` to see if it matches `"Darwin"`, seemed to be the best solution. If it's not possible or practical, I can accept the PR as it is.

**Comment by andresag01** [comment]
@sbutcher-arm:
> When I looked into if it was possible to detect OS X, inspecting CMAKE_SYSTEM_NAME to see if it matches "Darwin", seemed to be the best solution

There is a difference between what system CMake is compiling for (CMAKE_SYSTEM_NAME) and what system is running the build (CMAKE_HOST_SYSTEM_NAME). These should be set automatically, but the one we want for this specific change is CMAKE_SYSTEM_NAME as we need to modify the archiver flags only if we are compiling an Mbed TLS binary for OS X.

The APPLE variable should be automatically defined by CMake as well. However the docs for APPLE are confusing because they describe it differently depending on the CMake version you look at. Since this was unclear, I preferred to use CMAKE_SYSTEM_NAME in this PR, this should also help when later on the cross compilation problem is fixed (see [here](https://cmake.org/cmake/help/latest/variable/CMAKE_CROSSCOMPILING.html))

**Comment by sbutcher-arm** [comment]
@Patater - Could you please re-approve?

**Pull request merged**